### PR TITLE
First support for GIDS card (minidriver included by default in Windows)

### PR DESCRIPTION
**Pull request by vletoux**
GIDS stand for Generic Identity Device Specification.
It is a smart card specification made by Microsoft and which minidriver is included by default in Windows.
It can be found here: https://msdn.microsoft.com/en-us/library/windows/hardware/dn642100%28v=vs.85%29.aspx

This "driver" offers read/write support.
As a bonus, a compression routine has been added to the file compression.c

It has been extensively tested with:
- pkcs15-tool
- pkcs11-tool
- firefox with the keygen / certificate saving / TLS authentication with this page: https://openweb.or.kr/html5/e_keygen.php
- a third party pkcs11 test program

This driver can be used in parallele with the Microsoft minidriver.

To test it, a javacard applet has been developped and it is still in beta stage. It is not yet available but you can contact me to test it.

### CODE REVIEW

**Comment by frankmorgner** [comment]
This are quite a lot of changes. But if I understood correctly. you didn't have any independent test (i.e. some piece of hardware or software you didn't wrote yourself). Are there even any users for this kind of cards out there?

If I remember correctly, MS also introduced cardspace with a lot of noise not so long ago, but today it's almost forgotten. Will GIDS cards last longer?

**Comment by vletoux** [comment]
Why people are using PIV applet (except US gov) like the yubikey ? Because you don't have a driver to install and software to design. Why people are using the isoApplet ? Because you need to buy only basic javacard and not expensive middleware.
So yes, there is a demand for that, especially for SME for which it is difficult to use smart cards (complexity, cost, ...).
Check this blog article for more information:
https://randomoracle.wordpress.com/2013/03/31/piv-gids-home-brew-choosing-a-smart-card-standard-12/

About the design & tests, you are right : you have a chicken and egg problem here. 
But here is how I solved it:
- I created an applet based on the GIDS specification and the reverse of the Microsoft minidriver
(the card has been tested with the minidriver qualification program cmck.exe)
- I used the exactly same APDU sent by the minidriver to include them in the OpenSC driver

I documented everything I discovered here:
http://www.mysmartlogon.com/knowledge-base/generic-identity-device-specification-gids-kb/

I'm planning to release the applet card soon but during my tests I discovered one minor problem. I was using like the isoApplet a ram buffer to chain the APDU but you are then limiting maximum the file size to the memory size (<1kB). It is a problem when storing more than 10 keys because the cmapfile is growing after each new key.

The software stack (Microsoft minidriver, opensc pkcs11) can be then reused when needed to implement virtual smart card / token.
In my vision of the market, you'll have:
- cards deployed in large quantities (EID, PIV card, ...) => citizen
- GIDS card as a basic product (no pinpad support, single pin, ...) => for SME
- more customized cards with specific needs => large enterprise like banks for their employees

**GIDS is like PIV, but read write.**

Note: cardspace has nothing to do with smart card

**Comment by frankmorgner** [comment]
Since you tested your applet with the MS minidriver we at least have a good indication that your applet behaves conforming to the MS's standard. This gives us some indication that OpenSC uses the card as defined by the standard. That's good enough for me.

What is your opinion about the differences in specification and MS's minidriver behaviour listed in http://www.mysmartlogon.com/knowledge-base/generic-identity-device-specification-gids-kb/; what does this tell us about the status of standard and implementation?

But even if we would not have the standard, we could add *your* driver to support *your* applet. Please let us know when your applet or any other GIDS card is available. In the meantime we can leave this PR opened starting with the code review.

**Comment by vletoux** [comment]
You may have missed the GIDS specification that I referenced in my first post:
https://msdn.microsoft.com/en-us/library/windows/hardware/dn642100%28v=vs.85%29.aspx

Globally, you have the GIDS specification.
Then the data stored on the card (what Microsoft names electrical profile which is described in the annex of the minidriver specification v7.07 https://msdn.microsoft.com/en-us/library/windows/hardware/dn631754(v=vs.85).aspx)
Then some minidriver magic (like the compression algorithm of the certificates).

What is implemented in the minidriver is a subset of all of that.
For example, GIDS supports RSA from 1024 to 4096 and some ECC curves. In practice, the minidriver is limited to RSA 1024 & 2048.
GIDS support long (40 bytes) and short (8 bytes) challenge for admin authentication (mutual authentication versus general authentication). In practice, Windows 7 implement only the mutual authentication (40 bytes).
There is no APDU found in my driver that is not coming from the MS minidriver.

Because I wasn't able to configure ECC on the applet (because the minidriver doesn't use it), I'm unable to implement it in OpenSC.

Only one specific point differs from the specification. The minidriver SPEC and the GIDS spec specify 3DES ECB as the algorithm to computer the response of a challenge for the admin authentication.
BUT:
- the Gemalto minidriver tool use 3DES CBC (tested with a Gemalto ID Prime)
- the cmck.exe tool (for minidriver qualification) use also 3DES CBC
=> I decided to change 3DES ECB to 3DES CBC on my implementations

Ok for the PR. I wasn't expecting a commit before the applet was published and that's why I didn't send a mail on the OpenSC devel mailing list about that.
Expect a 1-2 week delay from now before publication.

**Comment by vletoux** [comment]
Applet published here:
https://github.com/vletoux/GidsApplet

File: src/libopensc/card-gids.h [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#ifdef _OPENSC_PKCS15_H
```
Comment (frankmorgner) [inline src/libopensc/card-gids.h]: If this header file depends on pkcs15.h, please include it here.

You could also add an include guard.

File: src/libopensc/card-gids.h [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+#define GIDS_MAX_CONTAINER 126
+
+// stolen from cardmod.h
+#define MAX_CONTAINER_NAME_LEN                  39
+#define CONTAINER_MAP_VALID_CONTAINER           1
+#define CONTAINER_MAP_DEFAULT_CONTAINER         2
+typedef struct _CONTAINER_MAP_RECORD
+{
+    unsigned short wszGuid [MAX_CONTAINER_NAME_LEN + 1];
```
Comment (frankmorgner) [inline src/libopensc/card-gids.h]: Please use tabs instead of spaces

File: src/libopensc/pkcs15-gids.c [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+
+#if HAVE_CONFIG_H
```
Comment (frankmorgner) [inline src/libopensc/pkcs15-gids.c]: please use ifdef

File: src/pkcs15init/gids.profile [hunk 4]
```diff
@@ -0,0 +1,11 @@
+#
+# PKCS15 r/w profile for SmartCard-HSM cards
```
Comment (frankmorgner) [inline src/pkcs15init/gids.profile]: copy/paste fault

**Comment by vletoux** [comment]
fixed the code according to your comments

File: src/libopensc/pkcs15-gids.c [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+
+	for (i = 0; i < recordsnum; i++) {
+		sc_cardctl_gids_get_container_t container;
+		memset(&container, 0, sizeof(sc_cardctl_gids_get_container_t));
+		container.containernum = i;
+		r = sc_card_ctl(card, SC_CARDCTL_GIDS_GET_CONTAINER_DETAIL, &container);
+		if (r < 0) {
+			break;
+		}
+		sc_pkcs15emu_gids_add_prkey(p15card, &container);
```
Comment (frankmorgner) [inline src/libopensc/pkcs15-gids.c]: you maybe want to do some error checking here.

File: src/pkcs15init/pkcs15-gids.c [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
```
Comment (frankmorgner) [inline src/pkcs15init/pkcs15-gids.c]: please add `#ifdef HAVE_CONFIG_H`

File: win32/OpenSC.wxs.in [hunk 7]
```diff
@@ -176,6 +176,9 @@
                 <Component Id="dnie_tool.exe" Guid="*" Win64="$(var.Win64YesNo)">
                   <File Source="$(var.SOURCE_DIR)\src\tools\dnie-tool.exe" Vital="yes"/>
                 </Component>
+				<Component Id="gids_tool.exe" Guid="*" Win64="$(var.Win64YesNo)">
```
Comment (frankmorgner) [inline win32/OpenSC.wxs.in]: this file uses spaces for indenting

**Comment by frankmorgner** [comment]
Please try to rebase onto OpenSC's master to remove your merge commit. Make sure to have a copy of your current branch if something fails!

**Comment by frankmorgner** [comment]
despite the minor comments above, this PR looks good. I'll wait some days longer if someone else wants to have a look

**Comment by vletoux** [comment]
I didn't see a remark about card-gids.c which have more than 1000 lines. Knowing you, that's something strange...

**Comment by frankmorgner** [comment]
:yum: Github hides files with more than 1k lines. I'll check later...

**Comment by frankmorgner** [comment]
- There is some needless casting in card-gids.c:
```int recordcount = (int) ((masterfilesize-1) / sizeof(gids_mf_record_t));```
This should be a `size_t`.
- This is also too much casting for my taste: `size = MIN((int) (data->buffersize - offset), (int) count);`
- There are a lot of magic values like `0xCB` or `0x5C`, which could use some more verbose naming via defines.
- I don't have a complete view about the minidriver, but doesn't OpenSC usually store the cardcf on the hard driver (instead of on the card like done with `gids_update_cardcf`)?
- `static int
gids_get_all_containers(sc_card_t* card, int *recordsnum)` should take an unsigned value as number, such as a `size_t`, same https://github.com/vletoux/OpenSC/blob/gids/src/libopensc/card-gids.c#L1021, https://github.com/vletoux/OpenSC/blob/gids/src/libopensc/card-gids.c#L1158, https://github.com/vletoux/OpenSC/blob/gids/src/libopensc/card-gids.c#L1159, https://github.com/vletoux/OpenSC/blob/gids/src/libopensc/card-gids.c#L1509
- https://github.com/vletoux/OpenSC/blob/gids/src/libopensc/card-gids.c#L1311-L1426 could use `sc_asn1_encode`, to be more readable.
- Please check if the crypto in `gids_authenticate_admin` could/should be realized via libsm

**Comment by frankmorgner** [comment]
Travis/AppVeyor are showing some compiler warnings:
```
gids-tool.c(475) : warning C4090: 'function' : different 'const' qualifiers
gids-tool.c:385:33: warning: format specifies type 'unsigned int' but the argument has type 'size_t' (aka 'unsigned long') [-Wformat]
gids-tool.c:385:33: warning: format specifies type 'unsigned int' but the argument has type 'size_t' (aka 'unsigned long') [-Wformat]
```

**Comment by vletoux** [comment]
gids-tool.c(475) : warning C4090: 'function' : different 'const' qualifiers
is linked to the argument parsing function hook used by Windows. You'll find it on every other tool.
=> I do not think it will be a good idea to fix the root problem with this PR

For the remark about  gids_authenticate_admin , the card do not support secure messaging. Have you something different in mind ?

For cardcf, yes, the OpenSC minidriver stores it in memory. But the Microsoft minidriver specification describe it as a file. If the file is not updated on the card itself, the cache used by the minidriver is not updated and the modification you made is not visible until a reboot of the computer (insert and remove is not enough)

Why  int recordcount should be a size_t ? Ok, the maximum records in less than 65000 so in practice it doesn't matter but I do not understand the logic behind. For me size_t represent the size of a buffer and usually the number of item is a int (or unsigned long).

The other remarks are ok for me

**Comment by frankmorgner** [comment]
- `gids-tool.c`: OK, I only copy/pasted from the log. However, it should be solved some day...
- `gids_authenticate_admin` uses 3DES with OpenSSL. You may use the wrappers provided by libsm (which is a misnomer if you ask me). Please check `sm-common.h`.
- changing cardcf to what is used by all other cards is an optional suggestion. I was hoping to reduce the code base by using a unified mechanism.
- the record number is an index, which is always 0 or positive. If I read the code correctly, you used `-1` only in case of errors. However, in OpenSC you have explicit error codes, so no need to overload parameters. See also https://stackoverflow.com/questions/3174850/type-for-array-index-in-c99 or https://matt.sh/howto-c. (Note that the size of `size_t` is architecture depended and not necessarily an `unsigned long`)

**Comment by vletoux** [comment]
Is that normal that sm-common.h is not referenced in any files except the libsm ones ?
I'm not confortable to include it because i will need to adjust the libopensc makefile to add libsm and the card which does not depend on OpenSSL except for the admin login will depend permanently on OpenSSL (because libsm cannot be compiled without OpenSSL)

**Comment by vletoux** [comment]
The magic value 0xCB is the GET DATA  INS code.
It is exactly the same code than https://github.com/OpenSC/OpenSC/blob/decd5ab987b6263bd855b683a271d10725dcf06e/src/libopensc/card-iasecc.c#L2839 or https://github.com/OpenSC/OpenSC/blob/451386886d208e0af9f8d4128696fc36cc9ad43a/src/libopensc/card-piv.c#L906

sc_asn1_encode could be used but the code is quite the same than https://github.com/OpenSC/OpenSC/blob/7120a9b549ba55f2ebac41300e129b9fe95dd454/src/libopensc/card-isoApplet.c#L763-L804 and this has already been tested a lot so I'm not very comfortable to change it.

Every other remarks have been fixed

**Comment by frankmorgner** [comment]
Errors have been done in the past. Turns out, you only get people to change their code as long as they have an incentive to do so. When code is merged this typically drops near zero so that the burdens of the past keep on haunting us forever.

So please use defines if possible (especially if you have the same value multiple times in your code). Using `sc_asn1_encode` would be nice to have since it is better readable. Also re-using existing OpenSSL code would be fine, though I have to admit that libsm would need some modifications to be applicable in the non-OpenSSL case (optional).

**Comment by philipWendland** [comment]
Hi,
if we are about to introduce defines for generic ISO 7816 constants, we
should do this in a way that allows other card drivers to re-use those.
Maybe in iso7816.h?

**Comment by vletoux** [comment]
code fixed according to comments

**Comment by frankmorgner** [comment]
- The `size_t`/`int` conversion could also be done in `gids_find_available_DO` and `gids_save_certificate`. In `gids_delete_cert` you forgot to remove the cast to `int`:
```
recordcount = (int) (masterfilebuffersize / sizeof(gids_mf_record_t));
```
- There are some lines that end with whitespaces, that easily be removed with a regex.
- Doesn't give casting an `int` to `u8` for creating `keyexchange`/`sign` in `gids_perform_create_keyfile` give a warning about possible dataloss? (same for `generatekey` in `gids_generate_key`)
- An optional suggestion: Instead of disabling the gids stuff when zlib is missing, you could rewrite the code to be more flexible. If I'm not mistaken this only means to rewrite `sc_compress` so that it throws an error if zlib is missing instead of not being implemented...

I think this is the last time I was going through the code. The overall quality looks very good. Last thing I want to do is to verify the functionality, i.e. loading this on a card and playing around a bit. Have you documented the lifecycle somewhere? (I can't find it on https://www.mysmartlogon.com/generic-identity-device-specification-gids-smart-card/)

**Comment by vletoux** [comment]
@frankmorgner What do you mean by lifecycle ?
The documented APDU to initialize a card after the applet installation or its termination ?
The APDU to initialize a card is described in the minidriver specification, annex GIDS electrical profile.
(chapter about the initialization of the masterfile doesn't give the exact APDU but just the data to put in a PUT DATA APDU)
The termination is described in the GIDS specification, TERMINATE APDU.
the gids-tool do initialize the card.

**Comment by frankmorgner** [comment]
Suppose I have  a card with an applet, how do I get keys, certificates and PINs on it from a "user" or "administrator" perspective? From your comment I assume that there is some Windows tool for this, but from the code I can see that `pkcs15-tool` might also do the trick...

**Comment by vletoux** [comment]
you set the PIN (user & admin) at initialization (sorry ...)
Then pkcs11-tool which generates / import key and OpenSC with the enginepkcs11 to generate certificate. Certificate can be saved using pkcs11-tool
pkcs15-tool can be used to dump objects.

**Comment by vletoux** [comment]
The easiest end-to-end test is https://openweb.or.kr/html5/e_keygen.php with the pkcs11 library.
Key generation, certificate request, certificate saving and certificate authentication, only one click away

**Comment by frankmorgner** [comment]
I tested and integrated your applet here https://github.com/frankmorgner/vsmartcard/commit/3fe38b0e6dd87711c4e4e0c2a2b751465640823e which mostly worked as expected. Unfortunately I did not manage to import the key...

1. generate key via https://openweb.or.kr/html5/e_keygen.php and export it as `$HOME/test.p12`
2. `gids-tool -X`
3. Now, neither `pkcs15-init --store-private-key $HOME/test.p12 --format pkcs12 --auth-id 80 --pin 1234` nor `pkcs11-tool --type privkey --write-object $HOME/test.p12` are working

I'm pretty sure I am doing something wrong (though I'd like to know what the problem exactly is). Anyway, when the above comments https://github.com/OpenSC/OpenSC/pull/651#issuecomment-181625630 are fixed I'm going to merge this PR if there are no more objections.

**Comment by vletoux** [comment]
I removed the int casting in `gids_find_available_DO` `gids_save_certificate` `gids_delete_cert`
I removed the whitespace at the end of the lines
I fixed the prototype of `gids_perform_create_keyfile` to replace int by u8.
`generatekey` var in `gids_generate_key` was already using u8 as external variable so this was not a problem (no compilation warning was given about `gids_perform_create_keyfile`)
GIDS could indeed be available without zlib, but given the fact that the certificates are compressed using zlib, this will be quite useless.

For the key import test, I was using the following code or directly with OpenSSL engine:
```
openssl.exe genpkey -algorithm RSA -out private_key.pem -pkeyopt rsa_keygen_bits:1024
pkcs11-tool --login --pin XXXX --write-object private_key.pem --type privkey 
```
This was related to my PR #669 (closed because of #655)
For the issue you are encountering, I think that the issue is related to the p12 file you are using. The code expect a private key file.
When I was talking about the website for the test, I was talking about testing the pkcs11 library and not to generate a p12 ... OpenSSL is made for that.

**Comment by frankmorgner** [comment]
I merged your commits by squashing the fixups and extending the commit message on the initial commit

**Comment by Jakuje** [comment]
Not sure if it is better to comment on this pull request or open new one. But two related things:

* Missing record in `.gitignore` lets git report untracked files: We should add line with `src/tools/gids-tool`.
* Missing documentation for `gids-tool`. It would be nice to have some manual page.

I can do the first one, but for the second one, I could do only copy-paste from usage (if it would be helpful), but documentation from authors would be nice.

**Comment by vletoux** [comment]
Indeed I forget to add the entry to .gitignore
What are you talking about missing documentation for gids-tool ?
I referenced the applet, the webpage on mysmartlogon.com and on the msdn site.
The best would be indeed to add a new issue and refer to this issue in any further pull request.

**Comment by Jakuje** [comment]
Sorry, I meant a manual page.  All the other tools have manual page in `doc/tools/`, except this one.

**Comment by frankmorgner** [comment]
@Jakuje we'd be happy to take a pull request!

**Pull request closed without merging**